### PR TITLE
Fix reverse translation for ap/hls instructions if result type > 64 bit

### PR DESCRIPTION
**Pull request by pauzinl**
This patch changes function call after reverse translation for ap/hls instructions with result type > 64 bit

### CODE REVIEW

File: lib/SPIRV/SPIRVReader.cpp [hunk 1]
```diff
@@ -2545,25 +2544,50 @@ CallInst *SPIRVToLLVM::transFixedPointInst(SPIRVInstruction *BI,
...
   IntegerType *Int1Ty = IntegerType::get(*Context, 1);
 
-  SmallVector<Type *, 7> ArgTys = {InTy,    Int1Ty,  Int32Ty,
-                                   Int32Ty, Int32Ty, Int32Ty};
-  FunctionType *FT = FunctionType::get(RetTy, ArgTys, false);
-
   Op OpCode = Inst->getOpCode();
   std::string FuncName =
       SPIRVFixedPointIntelMap::rmap(OpCode) + getFuncAPIntSuffix(RetTy, InTy);
+  auto Words = Inst->getOpWords();
```
[the remark this answers is not in the document]
Comment (pauzinl) [reply lib/SPIRV/SPIRVReader.cpp:2550]: I thought I should use less `if` statements here. If reducing code duplication is more priority, I will do it.

[the remark this answers is not in the document]
Comment (AlexeySotkin) [reply lib/SPIRV/SPIRVReader.cpp:2550]: I agree with @mlychkov, I think we should reduce code duplication.

[the remark this answers is not in the document]
Comment (pauzinl) [reply lib/SPIRV/SPIRVReader.cpp:2553]: Changed it in [1dd9bd3](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1254/commits/1dd9bd38c7ec7adee6f063ac5420899f4b24d7f4)
@AlexeySotkin is there any way to reduce extra instructions? Or this is acceptable? [(comment)](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1254#issuecomment-951996855)

File: lib/SPIRV/SPIRVReader.cpp [hunk 2]
```diff
@@ -2650,10 +2674,30 @@ CallInst *SPIRVToLLVM::transArbFloatInst(SPIRVInstruction *BI, BasicBlock *BB,
...
+    Value *Alloca = new AllocaInst(RetTy, SPIRAS_Private, "", BB);
+    Value *RetValPtr = new AddrSpaceCastInst(Alloca, RetPtrTy, "", BB);
+    std::vector<Value *> ArgsLocal = {
+        RetValPtr,
+        transValue(Inst->getOperand(0), BB->getParent(), BB) /* A - input */,
+        ConstantInt::get(Int32Ty,
+                         *WordsItr++) /* MA/Mout - width of mantissa */};
+    ArgTys = std::move(ArgTysLocal);
+    Args = std::move(ArgsLocal);
+  }
```
Comment (mlychkov) [inline lib/SPIRV/SPIRVReader.cpp:2700]:
```suggestion
  SmallVector<Type *, 8> ArgTys;
  std::vector<Value *> Args;

  if (RetTy->getIntegerBitWidth() > 64) {
    llvm::PointerType *RetPtrTy = llvm::PointerType::get(RetTy, SPIRAS_Generic);
    ArgTys.push_back(RetPtrTy);
    Value *Alloca = new AllocaInst(RetTy, SPIRAS_Private, "", BB);
    Value *RetValPtr = new AddrSpaceCastInst(Alloca, RetPtrTy, "", BB);
    ArgsLocal.push_back(RetValPtr);
  }
  ArgTys.insert(ArgTys.end(), {ATy, Int32Ty});
  // A - input
  ArgsLocal.emplace_back(transValue(Inst->getOperand(0), BB->getParent(), BB));
  // MA/Mout - width of mantissa
  ArgsLocal.emplace_back(ConstantInt::get(Int32Ty, *WordsItr++));
```

File: lib/SPIRV/SPIRVReader.cpp [hunk 3]
```diff
@@ -2677,17 +2721,37 @@ CallInst *SPIRVToLLVM::transArbFloatInst(SPIRVInstruction *BI, BasicBlock *BB,
...
+    auto *Func = cast<Function>(FCallee.getCallee());
+    Func->setCallingConv(CallingConv::SPIR_FUNC);
+    if (isFuncNoUnwind())
+      Func->addFnAttr(Attribute::NoUnwind);
+
+    return CallInst::Create(Func, Args, "", BB);
+  }
+
+  FunctionType *FT =
+      FunctionType::get(Type::getVoidTy(*Context), ArgTys, false);
```
Comment (mlychkov) [inline lib/SPIRV/SPIRVReader.cpp:2739]:
```suggestion
  Type *FuncRetTy =
      (RetTy->getIntegerBitWidth() <= 64) ? RetTy : Type::getVoidTy(*Context);
  FunctionType *FT = FunctionType::get(FuncRetTy, ArgTys, false);
```

File: lib/SPIRV/SPIRVReader.cpp [hunk 4]
```diff
@@ -2677,17 +2721,37 @@ CallInst *SPIRVToLLVM::transArbFloatInst(SPIRVInstruction *BI, BasicBlock *BB,
...
   FunctionCallee FCallee = M->getOrInsertFunction(FuncName, FT);
 
   auto *Func = cast<Function>(FCallee.getCallee());
   Func->setCallingConv(CallingConv::SPIR_FUNC);
   if (isFuncNoUnwind())
     Func->addFnAttr(Attribute::NoUnwind);
```
Comment (mlychkov) [inline lib/SPIRV/SPIRVReader.cpp:2746]:
```suggestion

  if (RetTy->getIntegerBitWidth() <= 64)
    return CallInst::Create(Func, Args, "", BB);
```

Comment (pauzinl) [reply lib/SPIRV/SPIRVReader.cpp:2722]: Thanks! Changed it in [5b525a8](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1254/commits/5b525a88aa6e1f1dd29eb71f2e8b493acc82837d)

**Review by MrSidims — APPROVED** [review]

**Comment by MrSidims** [comment]
@mlychkov were all your comments resolved?

**Review by mlychkov — APPROVED** [review]

**Comment by mlychkov** [comment]
> @mlychkov were all your comments resolved?

Yes, thanks a lot to @pauzinl for resolving and @MrSidims for pinging.

**Comment by MrSidims** [comment]
Ouch, I've 'rebase and merge'd, instead of 'squash and merge'd, sorry.

**Pull request merged**